### DISCUSSION TITLE
Allow declaring variables in `switch` block scope before the first label

### DISCUSSION THREAD

**Discussion by geekley** [dotnet/csharplang#7359] [Language Ideas]
Variables inside a `switch` block are scoped to the entire switch. This is fine because:
- If you want to limit scope to a case label, you can use a case block.
- You may not want case scope necessarily, e.g. if you're jumping through labels with `goto case` or if it's conceptually the same variable that you may or may not initialize and use depending on `case`.

The problem is that you're forced to declare the variable inside of a `case` label, even if you don't initialize it.
```cs
switch (a()) {
//  int x; // Why not allowed? It would make scope of x not confusing
case 0:
  f(0, y++); // x is not used
  break;
case 1:
  int x = 10; // CONFUSING! Makes it seem it's scoped to case 1 part, but it's not!
  f(x, y++);
  break;
case 2:
  f(x = -3, y++);
  goto case 1;
default:
  f(x = 0, y++);
  goto case 1;
}
```
We should be able to have declarations in a place that makes its scope clear to avoid confusion.
Currently, the only equivalent alternative that limits scope to switch is to add another block outside:
`{ int x; switch (a()) { ... } }` very ugly, and shouldn't be necessary.

There should be no impediments to implementing this if unassigned declarations run no code. It should be a simple grammar change with no (or few) implementation implications. It's equivalent to declaring it on first assignment, but cleaner.
Unless I'm missing something, like static constructor potentially being called (is it?).
Even if the declaration needs to run static ctor code, it can just run before all cases and after `a()` expression is evaluated (but before its value makes it jump to the corresponding label). It might be possible to use same logic to allow initializing before the first label too, but I don't have an opinion on whether this would be positive.

Uninitialized variable declarations between case labels should be permitted too - it's currently valid, but the compiler emits warnings CS0162 and IDE0035 thinking it's unreachable code. If it's not being initialized, it's not unreachable *code*, it's just a declaration.

**Comment by bernd5** [comment]
You could write the switch in a block with variable declarations like:

```csharp
using System;

int a()=> 1;
int f(int a, int b)=> 1;

int y = 123;
{
    int x;
    
    switch (a())
    {
        //  int x; // Why not allowed? It would make scope of x not confusing
        case 0:
          f(0, y++); // x is not used
          break;
        case 1:
          x = 10; // CONFUSING! Makes it seem it's scoped to case 1 part, but it's not!
          f(x, y++);
          break;
        case 2:
          f(x = -3, y++);
          goto case 1;
        default:
          f(x = 0, y++);
          goto case 1;
        }
}

```

**Reply by geekley** [reply]
Yes, I said that above.
> Currently, the only equivalent alternative that limits scope to switch is to add another block outside:
> `{ int x; switch (a()) { ... } }` very ugly, and shouldn't be necessary.

**Reply by bernd5** [reply]
Why is it ugly? I don't think so.
In my eyes it is ugly that variables are accessable across switch label sections - but that is how it is.

**Reply by geekley** [reply]
Ugly because it requires either completely unnecessary extra indentation or double `{{ }}` on the same line. You don't need this e.g. on `{ var ... ; for (...) {...} }` or `{var x ... ; using (x) {...} }` because you can `for (var ...) {...}` and `using (var ...) {...}`. Why waste the space that you could use for more succinct and clearer syntax?
The vars crossing case boundaries makes sense because a case label is just like a regular label.

The point of this proposal is not to "promote" this pattern, it's just to make the existing scope clearer, since we can't re-declare the variable with the same name in multiple cases without a `{...}` block. Specially considering declaring vars *between* case labels (after break, before next case) where it's not useful, won't give errors (just a warning), but it's not allowed before first label, where it would actually have the use of clarifying that scope spans across labels on the entire switch block.

**Reply by HaloFour** [reply]
Given the current behavior with `switch` is 21 years old in C#, and over 50 years old in C, it doesn't seem necessary to propose changes to clarify behavior that has always been there and that a developer only needs to learn once.  If anything it would be confusing to allow declarations there, but not arbitrary statements or expressions.  But allowing those things also wouldn't make any sense given the behavior of `switch` is intended to mimic a `goto` to a label within a block.
